src/firejail/dbus.c

```diff
@@ -18,41 +18,355 @@
+		sbox_exec_v(SBOX_USER | SBOX_SECCOMP | SBOX_CAPS_NONE | SBOX_KEEP_FDS, args);
+	} else {
+		if (close(status_pipe[1]) == -1 || close(args_pipe[0]) == -1)
+			errExit("close");
+
+		if (arg_dbus_user == DBUS_POLICY_FILTER) {
+			char *dbus_user_path_env = getenv(DBUS_SESSION_BUS_ADDRESS_ENV);
+			if (dbus_user_path_env == NULL) {
+				write_arg(args_pipe[1], DBUS_USER_SOCKET_PATH_FORMAT, getuid());
+			} else {
+				write_arg(args_pipe[1], dbus_user_path_env);
+			}
+			if (asprintf(&dbus_user_proxy_socket, DBUS_USER_PROXY_SOCKET_FORMAT,
+						 (int) getuid(), (int) getpid()) == -1)
+				errExit("asprintf");
+			write_arg(args_pipe[1], dbus_user_proxy_socket);
+			write_arg(args_pipe[1], "--filter");
+			write_profile(args_pipe[1], "dbus-user.");
+		}
+
+		if (arg_dbus_system == DBUS_POLICY_FILTER) {
+			write_arg(args_pipe[1], DBUS_SYSTEM_SOCKET_PATH);
+			if (asprintf(&dbus_system_proxy_socket, DBUS_SYSTEM_PROXY_SOCKET_FORMAT,
+						 (int) getuid(), (int) getpid()) == -1)
+				errExit("asprintf");
+			write_arg(args_pipe[1], dbus_system_proxy_socket);
+			write_arg(args_pipe[1], "--filter");
+			write_profile(args_pipe[1], "dbus-system.");
+		}
+
+		if (close(args_pipe[1]) == -1)
+			errExit("close");
+		char buf[1];
+		ssize_t read_bytes = read(status_pipe[0], buf, 1);
+		switch (read_bytes) {
+		case -1:
+			errExit("read");
+			break;
+		case 0:
+			fprintf(stderr, "xdg-dbus-proxy closed pipe unexpectedly\n");
+			// Wait for the subordinate process to write any errors to stderr and exit.
+			waitpid(dbus_proxy_pid, NULL, 0);
+			exit(-1);
+			break;
+		case 1:
+			if (arg_debug)
+				printf("xdg-dbus-proxy initialized\n");
+			break;
+		default:
+			assert(0);
+		}
 	}
+}
 
-	// blacklist the path
-	disable_file_or_dir(path);
-	free(path);
-	free(env_var);
+void dbus_proxy_stop(void) {
+	if (dbus_proxy_pid == 0)
+		return;
+	assert(dbus_proxy_status_fd >= 0);
+	if (close(dbus_proxy_status_fd) == -1)
+		errExit("close");
+	int status;
+	if (waitpid(dbus_proxy_pid, &status, 0) == -1)
+		errExit("waitpid");
+	if (WIFEXITED(status) && WEXITSTATUS(status) != 0)
+		fwarning("xdg-dbus-proxy returned %s\n", WEXITSTATUS(status));
+	dbus_proxy_pid = 0;
+	dbus_proxy_status_fd = -1;
+	if (dbus_user_proxy_socket != NULL) {
+		free(dbus_user_proxy_socket);
+		dbus_user_proxy_socket = NULL;
+	}
+	if (dbus_system_proxy_socket != NULL) {
+		free(dbus_system_proxy_socket);
+		dbus_system_proxy_socket = NULL;
+	}
+}
+
+static void socket_overlay(char *socket_path, char *proxy_path) {
+	int fd = safe_fd(proxy_path, O_PATH | O_NOFOLLOW | O_CLOEXEC);
+	if (fd == -1)
+		errExit("opening DBus proxy socket");
+	struct stat s;
+	if (fstat(fd, &s) == -1)
+		errExit("fstat");
+	if (!S_ISSOCK(s.st_mode)) {
+		errno = ENOTSOCK;
+		errExit("mounting DBus proxy socket");
+	}
+	char *proxy_fd_path;
+	if (asprintf(&proxy_fd_path, "/proc/self/fd/%d", fd) == -1)
+		errExit("asprintf");
+	if (mount(proxy_path, socket_path, NULL, MS_BIND | MS_REC, NULL) == -1)
```

Comment:
Would it be possible to instead set the `DBUS_SESSION_BUS_ADDRESS` environment variable to the proxy socket and blacklist the original session bus socket (or somehow avoid this mount on the session bus socket in another way)?

My corner case goes like this: Because some time passes before the mount is called, a user could replace `socket_path` with a link and mount `proxy_path` wherever he wants, including to places without search or write permission. Then a second race starts to control xdg-dbus-proxy by means of ptrace and friends. This possibly leads to issues if a tool with file capabilities or the setuid bit set has a socket path hard coded that it reasonably expects to be root controlled.

Reply:
Clearly, I wasn't at my best when writing this, since I mean to mount `proxy_fd_path` to `socket_path`, of course, not `proxy_path`. But your point still stands eminently: the user can mount `proxy_`(`fd_`)`path` wherever she wishes, if `socket_path` is a symbolic link. This isn't even that racy, because the proxy will only try to connect to `socket_path` one a connection to the proxy socket is made -- that's plenty of time to replace the link with a real socket. Even without ptrace, this could be used to effectively take over the system DBus (at least within the sandbox).

Updating `DBUS_SESSION_BUS_ADDRESS` (and hoping no program has hardcoded `/run/user/<uid>/bus`) is probably the best idea.

Reply:
> By the way, don't we have the same problem with the X11 socket?

I think it's a little bit different, because the mount target is a file we create ourselves in an empty tmpfs, and this tmpfs doesn't exist in the parent mount namespace. The implicit assumption is that no unprivileged process has access to the new sandbox mount namespace at this timepoint. If this assumption turns out to be wrong, the X11 socket whitelisting has exactly the same problem, yes.

As an additional hardening measure we could consider removing the `fchown` call:
https://github.com/netblue30/firejail/blob/c96efe7d0af04414ab3c9cd1256afd6ee678ff95/src/firejail/x11.c#L1308
This would mean the mount target is always owned by root, and in combination with the sticky bit on `/tmp/.X11-unix` that no unprivileged user can play with it.

Reply:
> By the way, don't we have the same problem with the X11 socket?

I'm going to prepare a patch to remove that attack surface

Reply:
Sorry for not getting back to this earlier, COVID-19 lockdown had added some extra tasks to my to-do list. Anyways, I realized that we _must_ bind-mount the proxy socket somewhere, so that we can bind-mount any empty directory over `/run/firejail/dbus/`. Otherwise, sandboxed apps will still be able to access the proxy sockets of other apps, which may have more permissive filters. So I guess one way to handle this is:
1. Bind-mount an empty file over the original `DBUS_SESSION_BUS_ADDRESS`
2. Bind-mount the session bus proxy socket to somewhere in `/run/firejail/mnt/` and set `DBUS_SESSION_BUS_ADDRESS accordingly.
3. Bind-mound the system bus socket to its usual place. This is safe, because `/run/dbus/system_bus_socket` is owned by root.
4. Bind-mount an empty directory over `/run/firejail/dbus/`.

Out of these, step 2 may still be problematic and suffer from the same issue as the X11 socket. However, the target of the bind-mount at step 2 may be owned by root, because only the proxy socket itself has to be owned by the user (to allow `xdg-dbus-proxy` clean up after itself when exiting). The bind-mount will be cleaned up without intervention from a user process when the mount namespace is destroyed.

Reply:
I think that sounds good.